typescript/cli/src/deploy/core.ts

```diff
@@ -0,0 +1,484 @@
+
+async function runChainSelectionStep(customChains: ChainMap<ChainMetadata>) {
+  const chainsToChoices = (chains: ChainMetadata[]) =>
+    chains.map((c) => ({ name: c.name, value: c.name }));
+  const choices: Parameters<typeof select>['0']['choices'] = [
+    new Separator('--Custom Chains--'),
+    ...chainsToChoices(Object.values(customChains)),
+    { name: '(New custom chain)', value: '__new__' },
+    new Separator('--Mainnet Chains--'),
+    ...chainsToChoices(mainnetChainsMetadata),
+    new Separator('--Testnet Chains--'),
+    ...chainsToChoices(testnetChainsMetadata),
+  ];
+
+  const local = (await select({
+    message: 'Select local chain (the chain to which you will deploy now)',
+    choices,
+    pageSize: 20,
+  })) as string;
+  handleNewChain([local]);
+
+  const remotes = (await checkbox({
+    message: 'Select remote chains the local will send messages to',
+    choices,
+    pageSize: 20,
+  })) as string[];
+  handleNewChain(remotes);
+  if (!remotes?.length) throw new Error('No remote chains selected');
+
+  const allChains = [local, ...remotes];
+  return { local, remotes, allChains };
+}
+
+async function runArtifactStep(allChains: ChainName[]) {
+  logBlue(
+    '\nDeployments can be totally new or can use some existing contract addresses.',
+  );
+  const isResume = await confirm({
+    message: 'Do you want use some existing contract addresses?',
+  });
+  if (!isResume) return undefined;
+
+  const artifactsPath = await input({
+    message: 'Enter filepath with existing contract artifacts (addresses)',
+  });
+  const artifacts = readDeploymentArtifacts(artifactsPath);
+  const artifactChains = Object.keys(artifacts).filter((c) =>
+    allChains.includes(c),
+  );
+  log(`Found existing artifacts for chains: ${artifactChains.join(', ')}`);
+  return artifacts;
+}
+
+async function runIsmStep(allChains: ChainName[]) {
+  logBlue(
+    '\nHyperlane instances requires an Interchain Security Module (ISM).',
+  );
+  const isMultisig = await confirm({
+    message: 'Do you want use a Multisig ISM?',
+  });
+  if (!isMultisig)
+    throw new Error(
+      'Sorry, only multisig ISMs are currently supported in the CLI',
+    );
+
+  const defaultConfigChains = Object.keys(defaultMultisigIsmConfigs);
+  const configRequired = !!allChains.find(
+    (c) => !defaultConfigChains.includes(c),
+  );
+  if (!configRequired) return;
+
+  logGray(
+    'Example config: https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/cli/typescript/cli/examples/multisig-ism.yaml',
+  );
+  const multisigConfigPath = await input({
+    message: 'Enter filepath for the multisig config',
+  });
+  const configs = readMultisigConfig(multisigConfigPath);
+  const multisigConfigChains = Object.keys(configs).filter((c) =>
+    allChains.includes(c),
+  );
+  log(`Found configs for chains: ${multisigConfigChains.join(', ')}`);
+  return configs;
+}
+
+function handleNewChain(chainNames: string[]) {
+  if (chainNames.includes('__new__')) {
```

Comment:
what is this?

Reply:
A special value to create a 'fake' option in the chain selection list for 'New Chain'. Adding a comment now

Reply:
In fact I'll pull this chain selection step code into another file for organization and reusability